File: ios/App/App/NativeSessionManager.swift
import Foundation
import FirebaseAuth

public class NativeSessionManager {
    static let shared = NativeSessionManager()
    
    private let uuidKey = "app_scoped_uuid"
    private let guestPhonePrefix = "guest_phone_"
    private let guestTimePrefix = "guest_time_"
    private let ttlMS: TimeInterval = 60 * 60 // 1 hour in seconds
    
    public func getTelemetryDeviceId() -> String {
        if let uuid = UserDefaults.standard.string(forKey: uuidKey) {
            return uuid
        }
        let newUuid = "ob-native-" + UUID().uuidString
        UserDefaults.standard.set(newUuid, forKey: uuidKey)
        return newUuid
    }
    
    public func cacheGuestPhone(orderId: String, phone: String) {
        UserDefaults.standard.set(phone, forKey: guestPhonePrefix + orderId)
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: guestTimePrefix + orderId)
    }
    
    public func getGuestPhone(orderId: String) -> String? {
        let time = UserDefaults.standard.double(forKey: guestTimePrefix + orderId)
        if Date().timeIntervalSince1970 - time > ttlMS {
            UserDefaults.standard.removeObject(forKey: guestPhonePrefix + orderId)
            UserDefaults.standard.removeObject(forKey: guestTimePrefix + orderId)
            return nil
        }
        return UserDefaults.standard.string(forKey: guestPhonePrefix + orderId)
    }
    
    public func getCurrentUserEmail() -> String? {
        return Auth.auth().currentUser?.email
    }
    
    public func hasValidSession(orderId: String) -> Bool {
        if Auth.auth().currentUser != nil {
            return true
        }
        return getGuestPhone(orderId: orderId) != nil
    }
}

File: android/app/src/main/java/com/bhojanos/orderbhojan/nativehost/NativeSessionManager.kt
package com.bhojanos.orderbhojan.nativehost

import android.content.Context
import android.content.SharedPreferences
import com.google.firebase.auth.FirebaseAuth
import java.util.UUID

object NativeSessionManager {
    private const val PREFS_NAME = "NativeSessionPrefs"
    private const val KEY_UUID = "app_scoped_uuid"
    private const val KEY_GUEST_PHONE_PREFIX = "guest_phone_"
    private const val KEY_GUEST_PHONE_TIME_PREFIX = "guest_time_"
    
    // 1 hour TTL for guest identity
    private const val TTL_MS = 60 * 60 * 1000L

    fun getTelemetryDeviceId(context: Context): String {
        // We use app-scoped UUID as fallback telemetry identifier.
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        var uuid = prefs.getString(KEY_UUID, null)
        if (uuid == null) {
            uuid = "ob-native-" + UUID.randomUUID().toString()
            prefs.edit().putString(KEY_UUID, uuid).apply()
        }
        return uuid
    }

    fun cacheGuestPhone(context: Context, orderId: String, phone: String) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit()
            .putString(KEY_GUEST_PHONE_PREFIX + orderId, phone)
            .putLong(KEY_GUEST_PHONE_TIME_PREFIX + orderId, System.currentTimeMillis())
            .apply()
    }

    fun getGuestPhone(context: Context, orderId: String): String? {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val time = prefs.getLong(KEY_GUEST_PHONE_TIME_PREFIX + orderId, 0)
        if (System.currentTimeMillis() - time > TTL_MS) {
            // Expired
            prefs.edit()
                .remove(KEY_GUEST_PHONE_PREFIX + orderId)
                .remove(KEY_GUEST_PHONE_TIME_PREFIX + orderId)
                .apply()
            return null
        }
        return prefs.getString(KEY_GUEST_PHONE_PREFIX + orderId, null)
    }

    fun getCurrentUserEmail(): String? {
        return FirebaseAuth.getInstance().currentUser?.email
    }

    fun hasValidSession(context: Context, orderId: String): Boolean {
        if (FirebaseAuth.getInstance().currentUser != null) {
            return true
        }
        return getGuestPhone(context, orderId) != null
    }
}
